app/Main.hs

```diff
@@ -99,44 +102,80 @@ pReleaseFormat = eitherReader $ \s ->
-            , taskRepository    = configRepo
-            , taskRevision      = configRevision
-            , taskReleaseFormat = releaseFormat }
-          rpath <- Hap.releasePath configDeployPath release
-          forM_ configCopyFiles $ \(C.CopyThing src dest) -> do
-            srcPath  <- resolveFile' src
-            destPath <- parseRelFile dest
-            let dpath = rpath </> destPath
-            (Hap.exec . Hap.MkDir . parent) dpath
-            Hap.scpFile srcPath dpath
-          forM_ configCopyDirs $ \(C.CopyThing src dest) -> do
-            srcPath  <- resolveDir' src
-            destPath <- parseRelDir dest
-            let dpath = rpath </> destPath
-            (Hap.exec . Hap.MkDir . parent) dpath
-            Hap.scpDir srcPath dpath
-          forM_ configBuildScript (Hap.playScript configDeployPath release)
-          Hap.registerReleaseAsComplete configDeployPath release
-          Hap.activateRelease configDeployPath release
-          Hap.dropOldReleases configDeployPath n
-          forM_ configRestartCommand Hap.exec
-        Rollback n -> do
-          Hap.rollback configDeployPath n
-          forM_ configRestartCommand Hap.exec
+    Right C.Config {..} -> do
+      chan <- newTChanIO
+      let printFnc dest str = atomically $
+            writeTChan chan (PrintMsg dest str)
+          hap sshOpts =  do
+            r <- Hap.runHapistrano sshOpts printFnc $
+              case optsCommand of
+                Deploy releaseFormat n -> do
+                  release <- Hap.pushRelease Task
+                    { taskDeployPath    = configDeployPath
+                    , taskRepository    = configRepo
+                    , taskRevision      = configRevision
+                    , taskReleaseFormat = releaseFormat }
+                  rpath <- Hap.releasePath configDeployPath release
+                  forM_ configCopyFiles $ \(C.CopyThing src dest) -> do
+                    srcPath  <- resolveFile' src
+                    destPath <- parseRelFile dest
+                    let dpath = rpath </> destPath
+                    (Hap.exec . Hap.MkDir . parent) dpath
+                    Hap.scpFile srcPath dpath
+                  forM_ configCopyDirs $ \(C.CopyThing src dest) -> do
+                    srcPath  <- resolveDir' src
+                    destPath <- parseRelDir dest
+                    let dpath = rpath </> destPath
+                    (Hap.exec . Hap.MkDir . parent) dpath
+                    Hap.scpDir srcPath dpath
+                  forM_ configBuildScript (Hap.playScript configDeployPath release)
+                  Hap.registerReleaseAsComplete configDeployPath release
+                  Hap.activateRelease configDeployPath release
+                  Hap.dropOldReleases configDeployPath n
+                  forM_ configRestartCommand Hap.exec
+                Rollback n -> do
+                  Hap.rollback configDeployPath n
+                  forM_ configRestartCommand Hap.exec
```

Comment:
@mrkkrp I think is worth it to have at least one end to end test for this new feature.